### PR TITLE
MGMT-8938: Add support for bootstrap-in-place day-2 worker flow

### PR DESCRIPTION
**Pull request by omertuc**
Note: These changes are for tests of the BIP flow in this repo
and they don't have anything to do with the Assisted Installer.

# Description

This commit adds a new Makefile target `deploy_ibip_worker` which runs a
new pytest test called `test_bip_add_worker`.

This new pytest "test" is identical to the existing
`test_bootstrap_in_place` test except once the installation of the BIP
SNO is done, it also adds a worker to that SNO.

Once the test starts the worker, it also waits for its CSRs to show up
and then approves them.

This test is made possible due to 4.11 support of adding workers to SNO.
This new test will not work on <=4.10.

# Background

Note that the new worker shouldn't boot a live-ISO with the
installer-generated `worker.ign` directly. Instead, it has to boot a
blank RHCOS Live ISO, then manually run `coreos-installer install...`
with the `worker.ign` file to install RHCOS with the `worker.ign`.

To achieve this in this test, we create *another* ignition file out
of a template in `src/assisted_test_infra/resources/sno-worker-live.ign.j2`,
and in this ignition file we:

- Copy the `worker.ign` file into the live filesystem
- Create a one-shot systemd unit service which runs coreos-installer with that worker.ign
- Reboot the node once RHCOS has been installed to disk

The worker then starts with the installed RHCOS+worker.ign and attempts to join the cluster

# Notable changes

- Added a new Makefile target to run the new flow
- Made the "running" terraform variable configurable. This allows me to
  create both the worker and the SNO and then only later control when they start.
- Added new Terraform config variable called `worker_image_path` which
  controls the ISO image used for workers. By default, if unset, it
  defaults to using the same image as the masters (this is the behavior
  all users of the terraform controller depend on).

### CODE REVIEW

File: src/tests/test_bootstrap_in_place.py
Comment (eranco74) [inline src/tests/test_bootstrap_in_place.py:386]: I'm guessing you choose this flow instead of embedding the worker.ign to RHCOS (not live ISO) because we already have the live ISO locally? because it's closer to how the assisted installer will do it?
Consider stating why we add the worker live-ISO since it's not that obvious

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eranco74) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:205]: Cosinder defining a const for oc path or oc command template

File: src/tests/test_bootstrap_in_place.py
Comment (eranco74) [inline src/tests/test_bootstrap_in_place.py:306]: In case this fails due to FileNotFoundError, log that this file is generated in 4.11 release and above

File: src/tests/test_bootstrap_in_place.py
Comment (eranco74) [inline src/tests/test_bootstrap_in_place.py:340]: Unsure what is going on here, you get the master IP for collecting logs? why here?

File: src/tests/test_bootstrap_in_place.py
Comment (eranco74) [inline src/tests/test_bootstrap_in_place.py:402]: It should take a lot less

File: src/tests/test_bootstrap_in_place.py
Comment (eranco74) [inline src/tests/test_bootstrap_in_place.py:399]: if this throws an exception you will not stop the approve_csr_worker thread,
consider adding try & finally

**Review by osherdp — CHANGES_REQUESTED** [review]

File: src/tests/test_bootstrap_in_place.py
Comment (omertuc) [reply src/tests/test_bootstrap_in_place.py:386]: Added a comment to clarify:
>       The reason we don't simply boot the live ISO with the worker.ign as
        ignition is because an RHCOS OS with worker.ign has to be written to
        disk, worker.ign is not meant to be the ignition of the live ISO
        itself. Basically, the live ISO phase is just a temporary operating
        system for the user to shell into and run coreos-installer to install
        the actual operating system used for OCP. In this test we just automate
        this manual process using our own ignition file, which will
        automatically run coreos-installer within the live operating system and
        reboot the node for us.

File: src/tests/test_bootstrap_in_place.py
Comment (omertuc) [reply src/tests/test_bootstrap_in_place.py:340]: Fixed `log_collection` function to look like this:

```python
    @JunitTestCase()
    def log_collection(self, master_ip: Optional[str]):
        """
        Collects all sorts of logs about the installation process

        @param master_ip The IP address of the master node. Used to SSH into the node when doing installer gather.
                         When not given, installer gather log collection is skipped.
        """
        ...
```

Now the worker phase will just pass master_ip as None to skip useless installer gather

File: Makefile
Comment (omertuc) [reply Makefile:249]: Oops didn't mean to commit the skipper removal

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (omertuc) [reply src/assisted_test_infra/test_infra/utils/oc_utils.py:224]: > maybe in this case we'd like to log warnings

I tried to keep the interface to these functions similar to the existing `get_operators_status` which I didn't want to modify. I guess it's best if I overhaul everything, including that function.

Comment (osherdp) [reply src/assisted_test_infra/test_infra/utils/oc_utils.py:224]: yeah I get you
it's just that we ignore lots of errors in test-infra, sometimes with no good reason

File: Makefile
Comment (omertuc) [reply Makefile]: Also updated corresponding to  openshift/release#28262

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:163]: Remove empty line

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:187]: Same here, add return value type hint annotation

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:202]:
```suggestion
    """
    Returns a dict with node names as keys and readiness as boolean values,
    e.g.: {"test-infra-cluster-master-0": True, "test-infra-cluster-worker-0": False}
    """
```

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:213]:
```suggestion
def get_unapproved_csr_names(kubeconfig: str) -> List[str]:
    """Returns a list of names of  all CertificateSigningRequest resources which are unapproved."""
```

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/utils/oc_utils.py:221]:
```suggestion
def approve_csr(kubeconfig: str, csr_name: str):
```

File: src/assisted_test_infra/test_infra/utils/oc_utils.py
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/utils/oc_utils.py:187]: Maybe add to the documentation that it raises SubprocessError

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:384]:
```suggestion
    @classmethod
    def approve_csrs(cls, kubeconfig: str, done: threading.Event):
```

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:356]:
```suggestion
    @classmethod
    def worker_ready(cls) -> bool:
```

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:309]:
```suggestion
    @classmethod
    def render_worker_live_iso_ignition(cls, install_device: str):
```

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:185]:
```suggestion
    @classmethod
    def all_operators_available(cls) -> bool:
```

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:226]: ssh_key `Expected type 'str', got 'Path' instead`

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:434]: 'image_path' value is not used

File: src/tests/test_bootstrap_in_place.py
Comment (eliorerz) [inline src/tests/test_bootstrap_in_place.py:412]: This method is a bit long, think about splitting it into few different methods

**Review by eliorerz — CHANGES_REQUESTED** [review]

File: src/assisted_test_infra/test_infra/helper_classes/config/base_entity_config.py
Comment (omertuc) [reply src/assisted_test_infra/test_infra/helper_classes/config/base_entity_config.py:20]: >It's not relevant only to this flow?

Not sure I understand this question

**Review by eliorerz — APPROVED** [review]
LGTM
/hold wait for e2e-metal-assisted-ipv4v6

**Comment by omertuc** [comment]
Waiting on #1654

**Comment by omertuc** [comment]
> > /test e2e-metal-assisted-ipv4v6
> 
> Not yet, you need image promotion of assisted-test-infra to finish

Oh I didn't realize it was merged so little ago. I just randomly checked it after hours, saw it was merged, and didn't realize it happened literally seconds after I checked

**Pull request merged**
